evennia/contrib/evmenu_login.py

```diff
@@ -0,0 +1,338 @@
+            string = dedent("""
+                |rYour password must be at least {} characters long.|n
+            """.strip("\n").format(LEN_PASSWD))
+            caller.msg(string)
+            # Loops on the same node
+            return create_password(caller, None)
+
+        # Everything's OK.  Create the new player account and
+        # possibly the character, depending on the multisession mode
+        from evennia.commands.default import unloggedin
+        # We make use of the helper functions from the default set here.
+        try:
+            permissions = settings.PERMISSION_PLAYER_DEFAULT
+            typeclass = settings.BASE_CHARACTER_TYPECLASS
+            new_player = unloggedin._create_player(caller, playername,
+                    password, permissions)
+            if new_player:
+                if settings.MULTISESSION_MODE < 2:
+                    default_home = ObjectDB.objects.get_id(
+                            settings.DEFAULT_HOME)
+                    unloggedin._create_character(caller, new_player, typeclass,
+                            default_home, permissions)
+        except Exception:
+            # We are in the middle between logged in and -not, so we have
+            # to handle tracebacks ourselves at this point. If we don't, we
+            # won't see any errors at all.
+            caller.msg(dedent("""
+                |rAn error occurred.|n  Please e-mail an admin if
+                the problem persists.
+            """.strip("\n")))
+            logger.log_trace()
+        else:
+            text = ""
+            caller.msg("Welcome, you're new account has been created!")
+            caller.sessionhandler.login(caller, new_player)
+
+    options = (
+        {
+            "key": "_default",
+            "desc": "Enter your new password.",
+            "goto": "create_password",
+        },
+    )
+    return text, options
+
+## Other functions
+
+def _formatter(nodetext, optionstext, caller=None):
+    """Do not display the options, only the text.
+
+    This function is used by EvMenu to format the text of nodes.
+    Options are not displayed for this menu, where it doesn't often
+    make much sense to do so.  Thus, only the node text is displayed.
+
+    """
+    return nodetext
+
+## Commands and CmdSets
+
+class UnloggedinCmdSet(CmdSet):
+    "Cmdset for the unloggedin state"
+    key = "DefaultUnloggedin"
+    priority = 0
+
+    def at_cmdset_creation(self):
+        "Called when cmdset is first created."
+        self.add(CmdUnloggedinLook())
+
+
+class CmdUnloggedinLook(Command):
+    """
+    An unloggedin version of the look command. This is called by the server
+    when the player first connects. It sets up the menu before handing off
+    to the menu's own look command.
+    """
+    key = syscmdkeys.CMD_LOGINSTART
+    locks = "cmd:all()"
+    arg_regex = r"^$"
+
+    def func(self):
+        "Execute the menu"
+        menu = EvMenu(self.caller, {
```

Comment:
True.  Since I had to define the functions and the MenuTree in the same file, however, I didn't know how to point to the same module.  It sounds like a good altrnative if defined in a separate module, however.